url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/examples/math/Vector3f.java

```diff
@@ -0,0 +1,389 @@
+     * @return A new normalized vector.
+     */
+    public static Vector3f normalize(Vector3f pv) {
+        float l = 1 / pv.length();
+        return new Vector3f(pv.x * l, pv.y * l, pv.z * l);
+    }
+    /**
+     * Inverts a given vector and stores the result into pvret.
+     * @param pv The source vector.
+     * @param pvret Inverted vector.
+     */
+    public static void invert(Vector3f pv, Vector3f pvret) {
+        pvret.x = -pv.x;
+        pvret.y = -pv.y;
+        pvret.z = -pv.z;
+    }
+    /**
+     * Inverts a given vector and returns a new inverted vector.
+     * @param pv The source vector.
+     * @return A new inverted vector.
+     */
+    public static Vector3f invert(Vector3f pv) {
+        return new Vector3f(-pv.x, -pv.y, -pv.z);
+    }
+    /**
+     * Returns the scalar product of two vectors.
+     * @param pv1 Left operand vector.
+     * @param pv2 Right operand vector.
+     * @return The scalar product.
+     */
+    public static float scalar(Vector3f pv1, Vector3f pv2) {
+        return (pv1.x * pv2.x) + (pv1.y * pv2.y) + (pv1.z * pv2.z);
+    }
+    /**
+     * Builds the cross product of two vectors an stores the result
+     * into pvret.
+     * @param pv1 Left operand vector.
+     * @param pv2 Right operand vector.
+	 * @param pvret The result vector.
+     */
+    public static void cross(Vector3f pv1, Vector3f pv2, Vector3f pvret) {
+        pvret.x = (pv1.y * pv2.z) - (pv1.z * pv2.y);
+        pvret.y = (pv1.z * pv2.x) - (pv1.x * pv2.z);
+        pvret.z = (pv1.x * pv2.y) - (pv1.y * pv2.x);
+    }
+    /**
+     * Returns the cross product of two vectors. 
+     * @param pv1 Left operand vector.
+     * @param pv2 Right operand vector.
+     * @return A new vector within the result of the operation.
+     */
+    public static Vector3f cross(Vector3f pv1, Vector3f pv2) {
+        return new Vector3f((pv1.y * pv2.z) - (pv1.z * pv2.y),
+                            (pv1.z * pv2.x) - (pv1.x * pv2.z),
+                            (pv1.x * pv2.y) - (pv1.y * pv2.x));
+    }
+    /**
+     * Returns the angle (phi) between two vectors.
+     * @param pv1 The first vector.
+     * @param pv2 The second vector.
+     * @return The angle.
+     */
+    public static float phi(Vector3f pv1, Vector3f pv2) {
+        return (float)Math.acos(scalar(pv1, pv2) / 
+                                (pv1.length() * pv2.length()));
+    }
+    /**
+     * Rotates a vector around the x-axis and stores the result into pvret.
+     * @param pv The vector which is to rotate.
+     * @param pangle The angle in rad of the rotation.
+     * @param pvret The return vector.
+     */    
+    public static void rotateX(Vector3f pv, float pangle, Vector3f pvret) {
+        float cosa = (float)Math.cos(pangle);
+        float sina = (float)Math.sin(pangle);
+        
+        pvret.x = pv.x;
+        pvret.y = (pv.y * cosa) + (pv.z * -sina);  
+        pvret.z = (pv.y * sina) + (pv.z * cosa);         
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 281:286, 306:311, 332:337.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.